src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
+
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
+    await confirmation;
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(3);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(3, null, eventData.data);
+    expect(model.data.confirmed).toEqual(eventData.data);
+  });
+
+  it<TestContext>('rejects the data in optimistic() and rolls back the changes if back-end rejects the data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    model.subscribe(subscriptionSpy);
+
+    await model.optimistic(eventData);
+    expect(model.data.optimistic).toEqual(eventData.data);
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-reject': true,
+        },
+      },
+    });
+
+    await subscriptionCalls[1];
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(2);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(2, null, eventData.data);
+    expect(model.data.confirmed).toEqual(undefined);
+  });
+
+  it<TestContext>('rejects the data and rolls back the changes if optimistic() timeouts', async ({
+    model,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    const [confirmation] = await model.optimistic(eventData, { timeout: 10 });
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    try {
+      await confirmation;
+    } catch (e) {
+      expect(e).toEqual(new Error('timed out waiting for event confirmation'));
+      expect(model.data.optimistic).toEqual(undefined);
+      expect(model.data.confirmed).toEqual(undefined);
+    }
+  });
+
+  it<TestContext>('rebases optimistic data on top of subsequent event data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    const channel = ably.channels.get(channelName);
+    const blazinglyFastEvent = {
```

Comment:
Not to be a party pooper, but this naming is probably confusing; there's nothing special about this even that makes it fast, so I'd prefer to just call it 'event' or something sensible :stuck_out_tongue: